lib/cn/cn_tree.c

```diff
@@ -1859,25 +1874,34 @@ cn_comp_cleanup(struct cn_compaction_work *w)
     if (w->cw_err) {
         if (!join) {
             struct kvset_list_entry *le = w->cw_mark;
+            struct cn_tree *tree = w->cw_tree;
 
             /* unmark input kvsets */
+            rmlock_wlock(&tree->ct_lock);
             for (uint kx = 0; kx < w->cw_kvset_cnt; kx++) {
-                assert(le);
+                assert(&le->le_link != &w->cw_node->tn_kvset_list);
 
                 /* A zspill may have been interrupted after moving kvsets to its znode. In which
                  * case cn_move() has already reset the workid in the kvsets.
                  */
-                assert(w->cw_action == CN_ACTION_ZSPILL || kvset_get_workid(le->le_kvset) != 0);
+                if (w->cw_action == CN_ACTION_ZSPILL) {
+                    if (kvset_get_work(le->le_kvset) == w)
+                        kvset_set_work(le->le_kvset, NULL);
```

Comment:
Why does zspill sometimes not have work set?

Reply:
@gsramdasi left a comment in his previous commit, zspill may have moved one or more, but not all kvsets from the root to the leaf and then get interrupted.  The kvsets that were moved by cn_move() will have had their work cleared.  But it's confusing to me, because wouldn't they also have been removed from the root node, and hence traversing it here is problematic???

Reply:
I don't see why we need a special handling for zspill here..

cn_move() doesn't leave a kvset in an intermediate state where the kvset is still in the root node but its workid is cleared. It either fails early without touching the kvsets in the root node or completes the move operation in which case it would have moved the kvsets from the root node to a target leaf node and reset their work ids.

Reply:
Yeah, that's what I thought, not sure why @gsramdasi added this...  @gsramdasi ???

Reply:
A zspill moves all kvsets to one child (znode), but it still needs to walk all children enqueuing/applying subspills. If the zspill thread receives an error and gets here, we could be here either before a subspill was applied to znode, or after the subspill was applied to znode, but the zspill wasn't finished with all the rest of the nodes yet.

If we get here before znode's subspill was applied, workid will be non-zero. If the subspill was applied (i.e. cn_move was called), the workid would be zero.

Reply:
While talking to @nabeelmmd about this, we realized there is a bug in the zspill path - https://jira.micron.com/jira/browse/NFSE-5383

Reply:
Another thing:  Is it possible for cw_mark to be invalid at this point?  In which case I should iterate over the node list instead of starting at cw_mark because there may no longer be cw_kvset_cnt kvsets in the node list.  Right @gsramdasi ?

Reply:
Yes, that's correct @beckerg. If cn_move has finished and set the workid to zero, the cw_mark kvset could be picked up for another compaction.

Reply:
I think it is better to leave this error handling to the zspill thread, like we do for cn_join().
The zspill thread is in a better position to know whether the error occurred before or after the znode's subspill was applied?